### PR TITLE
Add virtual block to speed up loading big files

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

Add virtual block concept to speed up loading big files. This is LOAD(server side) only and we need another client side change to read the correct page. 
Previously we would load metadata with data in one subtask, now we would have 2 subtasks to either load metadata or load data.

Remove Skipped file count since now it's hard to count them if we split into blocks, skipped byte information should be enough.
### Why are the changes needed?
Currently there's only one worker load one big file, which is slow and cause hot spot issue. So we introduce virtual block to split the big file into multiple partitions on different workers.
### Does this PR introduce any user facing changes?

"alluxio.dora.file.read.virtual.block.size"

### CODE REVIEW

**Comment by lucyge2022** [comment]
in this sense the read logic will need to be changed right?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 1]
```diff
@@ -520,34 +523,33 @@ public BlockWriter createFileWriter(String fileId, String ufsPath)
...
-  private boolean isAllPageCached(UfsFileStatus status, Optional<DoraMeta.FileStatus> fileStatus) {
-    if (!fileStatus.isPresent()) {
-      return false;
-    }
-    alluxio.grpc.FileInfo fi = fileStatus.get().getFileInfo();
-    if (!fi.getContentHash().equals(status.getContentHash())) {
-      return false;
+  private boolean isAllPageCached(Block block) {
+    alluxio.grpc.UfsStatus status = block.getUfsStatus();
+    String fileId = new AlluxioURI(status.getUfsFullPath()).hash();
```
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:528]: can u clean up all these places in pageddoraworker new AlluxioURI(ufsFullPath).hash()  to some common util places to generate file id?

**Comment by jja725** [comment]
> in this sense the read logic will need to be changed right?

yes, will be a PR for client side change as well

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -559,69 +561,79 @@ public ListenableFuture<LoadFileResponse> load(
...
+          skippedLength.addAndGet(block.getLength());
+          continue;
+        }
+        try {
+          ListenableFuture<Void> loadFuture = Futures.submit(() -> {
+            try {
+              if (options.hasUser()) {
+                AuthenticatedClientUser.set(options.getUser());
+              }
+              loadData(block.getUfsPath(), 0, block.getOffsetInFile(), block.getLength(),
```
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:588]: we dont need the mountid 0 any more do we?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 3]
```diff
@@ -559,69 +561,79 @@ public ListenableFuture<LoadFileResponse> load(
...
+                                    .setNoCache(false).setOffsetInFile(offset)
+                                    .setBlockSize(fileLength).build();
     String fileId = new AlluxioURI(ufsPath).hash();
-    ByteBuf buf = PooledDirectNioByteBuf.allocate((int) (4 * mPageSize));
-    try (BlockReader fileReader = createFileReader(fileId, 0, false, options)) {
-      // cache file data
-      while (fileReader.transferTo(buf) != -1) {
+    long bufferSize = 4 * mPageSize;
+    ByteBuf buf =
+        PooledByteBufAllocator.DEFAULT.directBuffer((int) Math.min(bufferSize, lengthToLoad));
```
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:627]: make it min of bufferSize or 1 page size if lengthToLoad is small, so these aligned buffers can be reused and reduce  framentation.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 4]
```diff
@@ -559,69 +561,79 @@ public ListenableFuture<LoadFileResponse> load(
...
+    try (BlockReader fileReader = createFileReader(fileId, offset, false, options)) {
```
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:628]: if positionShort is not use in that particular file can u help remove it?

Comment (lucyge2022) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:628]: actually loadData is only getting called by doraloadjob is it? why don't u just do mCacheManager.getAndLoad() and give it an empty buffer. so we dont waster extra buffer

File: common/transport/src/main/proto/grpc/block_worker.proto [hunk 5]
```diff
@@ -241,8 +242,8 @@ enum TaskStatus {
 message LoadFileRequest {
   repeated UfsStatus ufs_status = 1;
   required UfsReadOptions options = 2;
-  optional bool load_metadata_only = 3; // If set false, only the metadata of file will be loaded.
```
Comment (elega) [inline common/transport/src/main/proto/grpc/block_worker.proto:244]: why is this removed... i think zhihu will need to use this

File: dora/core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 6]
```diff
@@ -7100,6 +7100,16 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
           .setScope(Scope.WORKER)
           .build();
+  public static final PropertyKey DORA_READ_VIRTUAL_BLOCK_SIZE =
+      dataSizeBuilder(Name.DORA_READ_VIRTUAL_BLOCK_SIZE)
+          .setDefaultValue("0MB")
```
Comment (elega) [inline dora/core/common/src/main/java/alluxio/conf/PropertyKey.java:7105]: can you elaborate what will happen is this property is set to 0mb? does it mean this feature is disabled?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 7]
```diff
@@ -121,17 +117,21 @@ public class DoraLoadJob extends AbstractJob<DoraLoadJob.DoraLoadTask> {
   private final AtomicLong mTotalFailureCount = new AtomicLong();
   private final AtomicLong mCurrentFailureCount = new AtomicLong();
   private Optional<AlluxioRuntimeException> mFailedReason = Optional.empty();
-  private Iterator<UfsStatus> mUfsStatusIterator;
+  private final Iterable<UfsStatus> mUfsStatusIterable;
+  private Optional<Iterator<UfsStatus>> mUfsStatusIterator = Optional.empty();
```
[the remark this answers is not in the document]
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:120]: under which circumstances can this iterator be null?

**Comment by elega** [comment]
i'm just thinking whether the implementation will be easier if we just represent virtual blocks of large files implicitly, e.g. just name the vb1 of the file "foo" to be "foo_1", and vb2 "foo_2". There can be a field in the metadata indicating this is a virtual block instead of a file. This way we don't need to change the hashing algorithm and the client side implementation should be easier.

Also can we prioritize the client side change first as not everyone uses the distributed load tool to load the data.

Thanks!

File: common/transport/src/main/proto/grpc/block_worker.proto (hunk 5 above)
Comment (jja725) [reply common/transport/src/main/proto/grpc/block_worker.proto:244]: now we seperate the loadmetadata & load data task so if we specify ufsStatus and don't specify blocks then we only load metadata so this flag is no longer useful.

**Comment by jja725** [comment]
> i'm just thinking whether the implementation will be easier if we just represent virtual blocks of large files implicitly, e.g. just name the vb1 of the file "foo" to be "foo_1", and vb2 "foo_2". There can be a field in the metadata indicating this is a virtual block instead of a file. This way we don't need to change the hashing algorithm and the client side implementation should be easier.
> 
> also can we prioritize the client side change first as not everyone uses the distributed load tool to load the data.

That's a good point, I think the major difference between implicit/explicit representation of virtual block is whether we want to include this in our metadata. I'm open to discussion. Right now I just keep the original metadata format for simplicity since I didn't find a proper usage of that virtual block information

File: dora/core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 8]
```diff
@@ -1145,15 +1145,15 @@ public static String getSyncMetricName(long mountId) {
           .setMetricType(MetricType.COUNTER)
           .build();
 
-  public static final MetricKey MASTER_JOB_LOAD_FILE_COUNT =
-      new Builder("Master.JobLoadFileCount")
-          .setDescription("The number of files loaded by load commands")
+  public static final MetricKey MASTER_JOB_LOAD_INODES_COUNT =
+      new Builder("Master.JobLoadInodesCount")
+          .setDescription("The number of files/dirs loaded by load commands")
```
[the remark this answers is not in the document]
Comment (jja725) [reply dora/core/common/src/main/java/alluxio/metrics/MetricKey.java:1150]: yes, UfsStatusIterator would give us dir metadata

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jja725) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:120]: Previously we have verify functionality so we need iterable. remove iterable to avoid confusion.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 1 above)
Comment (jja725) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:528]: I would defer this to the PR replacing AlluxioURI with UFSURI

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
Comment (jja725) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:588]: probably we need to integrate this with multi mount point change so we still need to input something later on? @dbw9580

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 4 above)
Comment (jja725) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:628]: Ideally we should have dedicate function to load from UFS and put into cacheManager and don't return anything. Would do in a later PR.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 3 above)
Comment (jja725) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:627]: done

**Comment by jja725** [comment]
@dbw9580 ConsistentHashProviderTest.workerListTtl is failing consistently in my PR, but there's no problem when I run locally, do you have any guess what's the problem?

**Comment by dbw9580** [comment]
> @dbw9580 ConsistentHashProviderTest.workerListTtl is failing consistently in my PR, but there's no problem when I run locally, do you have any guess what's the problem?

```
[WARNING] Flakes: 
2023-10-10T00:54:31.9569293Z 00:54:31.956 [WARNING] alluxio.client.file.dora.ConsistentHashProviderTest.workerListTtl
2023-10-10T00:54:31.9569938Z 00:54:31.956 [ERROR]   Run 1: ConsistentHashProviderTest.workerListTtl:210 expected:<0> but was:<1>
2023-10-10T00:54:31.9570544Z 00:54:31.956 [INFO]   Run 2: PASS
```

It's a bit flaky. Can you try this patch to see if it can help reduce the flakiness?

```patch
Index: dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashProviderTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashProviderTest.java b/dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashProviderTest.java
--- a/dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashProviderTest.java	(revision ba61600f3215f3340b3c808c00352bc808ec2d8f)
+++ b/dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashProviderTest.java	(date 1696921029083)
@@ -215,7 +215,7 @@
         provider.getActiveNodesMap());
 
     // after TTL expires, refresh should change the worker list and the active nodes map
-    Thread.sleep(WORKER_LIST_TTL_MS);
+    Thread.sleep(2 * WORKER_LIST_TTL_MS);
     provider.refresh(newList, NUM_VIRTUAL_NODES);
     assertEquals(1, provider.getUpdateCount() - initialUpdateCount);
     assertEquals(newList, provider.getLastWorkerInfos());

```

File: dora/shell/src/main/java/alluxio/common/SelectWorkerKey.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.common;
+
+/**
+ * Interface for keys used to present an object which can be used with
+ * {@link alluxio.client.file.dora.WorkerLocationPolicy}.
+ */
+public interface SelectWorkerKey {
```
Comment (lucyge2022) [inline dora/shell/src/main/java/alluxio/common/SelectWorkerKey.java:18]: can u name it ShardKey or sth equivalent
and the func to be  asString()

Comment (jja725) [reply dora/shell/src/main/java/alluxio/common/SelectWorkerKey.java]: done

**Review by lucyge2022 — APPROVED** [review]
LGTM.

**Comment by jja725** [comment]
@huanghua78 I'm not sure about the release status for AI 3.0, bring this to your awareness since it's a big change

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Comment by jja725** [comment]
alluxio-bot, merge this please
